standardization/src/main/scala/za/co/absa/enceladus/standardization/interpreter/stages/TypeParser.scala

```diff
@@ -15,185 +15,331 @@
-      case t                => throw new IllegalStateException(s"${t.typeName} type cannot be cast to a strong typed format in this version of Enceladus")
-    }
-  }
 
-  /**
-    * Deals with specific weirdness that we see with dates & times coming from various tech stacks
-    * Also supports epoch format
-    */
-  private def dateTimePreProcess(field: StructField, origSchema: StructType, path: String, attr: Column, epochFormat: String, castFunction: (Column, String) => Column): Column = {
-    val format = DateTimePattern(field)
-
-    if (DateTimePattern.isEpoch(format)) {
-      from_unixtime(attr.cast("Long") / format.epochFactor, epochFormat)
-    } else {
-      val origType: Option[DataType] = SchemaUtils.getFieldType(path, origSchema) // sadly with parquet support, incoming might not be all `plain`
-      // this case covers some IBM date format where it's represented as a double ddmmyyyy.hhmmss
-      val sourceCol: Column = origType.get match {
-        case _: DoubleType | _: FloatType =>
-          logger.warn(s"$path is specified as timestamp or date, but original type is Double. Trying to interpret as string.")
-          if (format.isDefault) {
-            throw new InvalidParameterException("Dates & times represented as numeric values need specified 'pattern' metadata")
-          }
-          val Array(l, r) = format.split("\\.")
-          val precision = l.length() + r.length()
-          val scale = r.length()
-          attr.cast(DecimalType.apply(precision, scale)).cast(StringType)
-        case _: DecimalType =>
-          logger.warn(s"$path is specified as timestamp or date, but original type is Decimal. Trying to interpret as string.")
-          if (format.isDefault) {
-            throw new InvalidParameterException("Dates & times represented as numeric values need specified 'pattern' metadata")
-          }
-          attr.cast(StringType)
-        case _ => attr
-      }
-      castFunction(sourceCol, format)
-    }
+  def standardize(field: StructField, path: String, origSchema: StructType)
+                 (implicit spark: SparkSession, udfLib: UDFLibrary): ParseOutput = {
+    ParserWorker(field, path, origSchema).standardize
   }
 
-  private def toTimestamp(field: StructField, origSchema: StructType, path: String, attr: Column): Column = {
-    dateTimePreProcess(field: StructField, origSchema: StructType, path: String, attr: Column, epochTimestampFormat, to_timestamp)
+  private type Parent = Option[Either[Column, Column]]
+  private type ColumnTransformer = Column=>Column
+
+  private object Parent {
+    def array(column: Column): Parent = Some(Left(column))
+    def struct(column: Column): Parent = Some(Right(column))
   }
 
-  private def toDate(field: StructField, origSchema: StructType, path: String, attr: Column): Column = {
-    dateTimePreProcess(field: StructField, origSchema: StructType, path: String, attr: Column, epochDateFormat, to_date)
+  private sealed trait ParserWorker {
```

Comment:
True, but we could just as easily put all of our classes in the same file because they all stem from the `main` method needing them, directly or transitively. 
I'd suggest using `package` access modifier and separating things a bit.